Enhance "Custom dictionary style" sample to use DictionarySymbolStyle created from a web style (#624)
Esri/arcgis-runtime-samples-java: symbology/custom-dictionary-style/README.md

## How to use the sample

Use the radio buttons to toggle between the dictionary symbols from the web style and style file. Pan and zoom around the map to see the symbology from the chosen dictionary symbol style. The web style and style file are slightly different to each other to give a visual indication of the switch between the two.

## How it works

1. Create a `PortalItem`, referring to a `Portal` and the item ID of the web style.
2. Based on the style selected:
-  If the web style toggle has been selected, create a new `DictionarySymbolStyle` from the portal item using `new DictionarySymbolStyle(portalItem)`
-  If the file style toggle has been selected, create a new dictionary symbol style using `DictionarySymbolStyle.createFromFile(stylxFile.getAbsolutePath())`

> 3. If the file style toggle has been selected, create a new dictionary symbol style using `DictionarySymbolStyle.createFromFile(stylxFile.getAbsolutePath())`

Reviewer: Looks like you can delete point 3 here, and update the numbers further down afterwards 😄

## How to use the sample

> Use the radio buttons to toggle between the dictionary symbols from the web style and style file. Pan and zoom around the map to see the symbology from the chosen dictionary symbol style.

Reviewer 2: ```suggestion
Use the radio buttons to toggle between the dictionary symbols from the web style and style file. Pan and zoom around the map to see the symbology from the chosen dictionary symbol style. The web style and style file are slightly different to each other to give a visual indication of the switch between the two.
```

Reviewer 2: Bit of unsolicited review there :D. Not sure if the sentence I've add reads well, but just something to clue the user in to that expected difference when using one style vs the other

## How it works

1. Create a `PortalItem`, referring to a `Portal` and the item ID of the web style.

> 2. If the web style toggle has been selected, create a new `DictionarySymbolStyle` from the portal item using `new DictionarySymbolStyle(portalItem)`
> 3. If the file style toggle has been selected, create a new dictionary symbol style using `DictionarySymbolStyle.createFromFile(stylxFile.getAbsolutePath())`

Reviewer 2: ```suggestion
2. Based on the style selected:
  -  If the web style toggle has been selected, create a new `DictionarySymbolStyle` from the portal item using `new DictionarySymbolStyle(portalItem)`
  -  If the file style toggle has been selected, create a new dictionary symbol style using `DictionarySymbolStyle.createFromFile(stylxFile.getAbsolutePath())`
```

Just stepping out of the numbered ordering here since they're "ifs"

Reviewer: Agree with this. I also think we could perhaps incoroprate point 1 (creating the PortalItem`) into the respective 'if' here, as the portal item is only necessary for that particular action, not for the whole sample.

4. Load the symbol dictionary asynchronously using `dictionarySymbolStyle.loadAsync()` and wait for it to completely load using `dictionarySymbolStyle.addDoneLoadingListener()`.
5. Create a new `DictionaryRenderer`, providing the dictionary symbol style.
6. Apply the dictionary renderer to a feature layer using using `FeatureLayer.setRenderer(dictionaryRenderer)`.
7. Add the feature layer to map using `Map.getOperationalLayers().add(FeatureLayer)`.

## Relevant API

* DictionaryRenderer
* DictionarySymbolStyle
* Portal
* PortalItem

## About the data

> The data used in this sample is from a feature layer showing a subset of [restaurants in Redlands, CA](https://services2.arcgis.com/ZQgQTuoyBrtmoGdP/arcgis/rest/services/Redlands_Restaurants/FeatureServer) hosted as a feature service with attributes for rating, style, health score, and open hours. The feature layer is symbolized using a dictionary renderer that displays a single symbol for all of these variables. The renderer uses symbols from a custom restaurant dictionary style created from a [stylex file](https://arcgisruntime.maps.arcgis.com/home/item.html?id=751138a2e0844e06853522d54103222a) and a [web style](https://arcgisruntime.maps.arcgis.com/home/item.html?id=adee951477014ec68d7cf0ea0579c800), available as an items from ArcGIS Online, to show unique symbols based on several feature attributes. The symbols it contains were created using ArcGIS Pro. The logic used to apply the symbols comes from an Arcade script embedded in the stylx file (which is a SQLite database), along with a JSON string that defines expected attribute names and configuration properties.

Reviewer: Maybe we can break this down into two paragraphs, one about the data, one about the renderer?

```suggestion
The data used in this sample is from a feature layer showing a subset of [restaurants in Redlands, CA](https://services2.arcgis.com/ZQgQTuoyBrtmoGdP/arcgis/rest/services/Redlands_Restaurants/FeatureServer) hosted as a feature service with attributes for rating, style, health score, and open hours.

The feature layer is symbolized using a dictionary renderer that displays a single symbol for all of these variables. The renderer uses symbols from a custom restaurant dictionary style created from a [stylex file](https://arcgisruntime.maps.arcgis.com/home/item.html?id=751138a2e0844e06853522d54103222a) and a [web style](https://arcgisruntime.maps.arcgis.com/home/item.html?id=adee951477014ec68d7cf0ea0579c800), available as an items from ArcGIS Online, to show unique symbols based on several feature attributes. The symbols it contains were created using ArcGIS Pro. The logic used to apply the symbols comes from an Arcade script embedded in the stylx file (which is a SQLite database), along with a JSON string that defines expected attribute names and configuration properties.
```

## How to use the sample

Use the radio buttons to toggle between the dictionary symbols from the web style and style file. Pan and zoom around the map to see the symbology from the chosen dictionary symbol style. The web style and style file are slightly different to each other to give a visual indication of the switch between the two.

## How it works

1. Create a `PortalItem`, referring to a `Portal` and the item ID of the web style.
2. Based on the style selected:
-  If the web style toggle has been selected, create a new `DictionarySymbolStyle` from the portal item using `new DictionarySymbolStyle(portalItem)`

> -  If the file style toggle has been selected, create a new dictionary symbol style using `DictionarySymbolStyle.createFromFile(stylxFile.getAbsolutePath())`

Reviewer 3: ```suggestion
-  If the file style toggle has been selected, create a new `DictionarySymbolStyle` using `DictionarySymbolStyle.createFromFile(stylxFile.getAbsolutePath())`
```

> 3. Load the symbol dictionary asynchronously using `dictionarySymbolStyle.loadAsync()` and wait for it to completely load using `dictionarySymbolStyle.addDoneLoadingListener()`.

Reviewer 3: we can simplify this step a bit: keeping the focus on the dictionary symbol functionality without explicitly spelling out the loading API pattern. 
```suggestion
3. Load the dictionary symbol style. Once it has loaded, create a new `DictionaryRenderer`, providing the dictionary symbol style.
```

> 4. Create a new `DictionaryRenderer`, providing the dictionary symbol style.

Reviewer 3: we should have mention here of the other constructor too: the one with the field overrides (it's used for the web style option: any reason why and we can explain that to the user here?

Author: The field overrides are used so that the symbols created from the web style are different from the symbols automatically rendered from the style file. 
```suggestion
4. To contrast the symbols rendered from the two `DictionarySymbolStyle`s, create a new `DictionaryRenderer` providing the dictionary symbol style, and manually map the attributes from the feature layer to those expected by the `DictionarySymbolStyle`.
```

> 5. Apply the dictionary renderer to a feature layer using `FeatureLayer.setRenderer(dictionaryRenderer)`.

Reviewer 3: ```suggestion
5. Apply the dictionary renderer to a feature layer using `featureLayer.setRenderer(dictionaryRenderer)`.
```

> 6. Add the feature layer to map using `Map.getOperationalLayers().add(FeatureLayer)`.

Reviewer 3: ```suggestion
6. Add the feature layer to the map's operational layers using `getOperationalLayers().add(featureLayer)`.
```

## How it works

1. Create a `PortalItem`, referring to a `Portal` and the item ID of the web style.
2. Based on the style selected:
-  If the web style toggle has been selected, create a new `DictionarySymbolStyle` from the portal item using `new DictionarySymbolStyle(portalItem)`, and load it
-  If the file style toggle has been selected, create a new `DictionarySymbolStyle` using `DictionarySymbolStyle.createFromFile(stylxFile.getAbsolutePath())`
3. Create a new `DictionaryRenderer`, providing the dictionary symbol style.
4. Apply the dictionary renderer to a feature layer using `featureLayer.setRenderer(dictionaryRenderer)`.
5. Add the feature layer to the map's operational layers using `getOperationalLayers().add(featureLayer)`.

## Relevant API

* DictionaryRenderer
* DictionarySymbolStyle
* Portal
* PortalItem

## About the data

The data used in this sample is from a feature layer showing a subset of [restaurants in Redlands, CA](https://services2.arcgis.com/ZQgQTuoyBrtmoGdP/arcgis/rest/services/Redlands_Restaurants/FeatureServer) hosted as a feature service with attributes for rating, style, health score, and open hours.

> The feature layer is symbolized using a dictionary renderer that displays a single symbol for all of these variables. The renderer uses symbols from a custom restaurant dictionary style created from a [stylex file](https://arcgisruntime.maps.arcgis.com/home/item.html?id=751138a2e0844e06853522d54103222a) and a [web style](https://arcgisruntime.maps.arcgis.com/home/item.html?id=adee951477014ec68d7cf0ea0579c800), available as an items from ArcGIS Online, to show unique symbols based on several feature attributes. The symbols it contains were created using ArcGIS Pro. The logic used to apply the symbols comes from an Arcade script embedded in the stylx file (which is a SQLite database), along with a JSON string that defines expected attribute names and configuration properties.

Reviewer 4: ```suggestion
The feature layer is symbolized using a dictionary renderer that displays a single symbol for all of these variables. The renderer uses symbols from a custom restaurant dictionary style created from a [stylx file](https://arcgis.com/home/item.html?id=751138a2e0844e06853522d54103222a) and a [web style](https://arcgis.com/home/item.html?id=adee951477014ec68d7cf0ea0579c800), available as an items from ArcGIS Online, to show unique symbols based on several feature attributes. The symbols it contains were created using ArcGIS Pro. The logic used to apply the symbols comes from an Arcade script embedded in the stylx file (which is a SQLite database), along with a JSON string that defines expected attribute names and configuration properties.
```

Reviewer 4: Ha I thought I was on devtopia... Will make the same suggestion there! 😅
